File: drawing-hg/meltplot/MeltParser.py
#! /usr/bin/python

# Python script to take input from command line and process .pairs and
# .fpairs files to make x-y data for plotting a melt curve.
#
# For the example of a multi stranded melt (output from
# concentrations) with a melt range from 10 to 40 deg. C in 10 degree
# increments, the input files are:
#   ./10.0/prefix.fpairs
#   ./20.0/prefix.fpairs
#   ./30.0/prefix.fpairs
#   ./40.0/prefix.fpairs
#
# The usage is:
#  MeltParser.py filename tempStart tempIncrement tempEnd outputfile
#
#  filename is the name of the files, e.g. prefix.ppairs or
#    prefix.fpairs.  The files must be of the format as outputted
#    by the programs pairs and concentrations, respectively.
#  tempStart is the starting temperature for the melt profile
#  tempIncrement is the temperature increment for the melt profile
#  tempEnd is the ending temperature
#  outputfile is the name of the output file to which the data are
#   written.  This file will in turn be read by MeltPlot.Py to generate
#    the plot.
#
# January, 2007, Justin Bois

# Import the modules we'll use
import sys
argv = sys.argv
exit = sys.exit
import re
WhiteSpaceSearch = re.compile('\s+')


# ######### GET DATA FROM COMMAND LINE INPUT ##################
filename = argv[1]
tempStart = float(argv[2])
tempIncrement = float(argv[3])
tempEnd = float(argv[4])
outputFile = argv[5]

if filename[-6:] == 'ppairs':
    ppairsMode = 1
else:
    ppairsMode = 0

Tlist = []
temp = tempStart
while temp <= tempEnd + 0.0000001: # Add small number to avoid rounding error
    Tlist.append(temp)
    temp += tempIncrement

fUnpaired = [0.0]*len(Tlist)

if ppairsMode:
    for k in range(len(Tlist)):
        inputFile = './%.1f/%s' % (Tlist[k],filename)
        f = open(inputFile)

        # Now blow through lines to number of bases
        line = f.readline()
        while not(line[0].isdigit()) and line != '':
            line = f.readline()
        if line == '':
            print 'Error: No records found in input file %s.' % inputFile
            exit(1)

        # The line we're at now is the total number of bases
        LineList = WhiteSpaceSearch.split(line)
        nBases = int(LineList[0])

        # Now read records and find unpaired bases
        line = f.readline()
        while line != '' and line[0].isdigit():
            LineList = WhiteSpaceSearch.split(line)
            if int(LineList[1]) == nBases + 1:
                fUnpaired[k] += float(LineList[2])
            line = f.readline()
        f.close()

        fUnpaired[k] /= nBases

else: # fpairs file
    for k in range(len(Tlist)):
        inputFile = './%.1f/%s' % (Tlist[k],filename)
        f = open(inputFile)

        # Blow through comment lines until we get to inital monomer concentrations
        line = f.readline()
        while line[0:17] != '% Initial monomer' and line != '':
            line = f.readline()
        if line == '':
            print 'Error: No initial monomer concentrations found in %s.' % inputFile
            exit(1)

        line = f.readline()
        # We are now at the first initial monomer concentration
        x0 = []
        while (len(line) > 10 and line[4].isdigit()):
            LineList = WhiteSpaceSearch.split(line)
            x0.append(float(LineList[2]))
            line = f.readline()

        # We have passed the initial monomer concentrations.  Now read to sequences
        Nstrands = len(x0)
        while line[0:13] != '% id sequence' and line != '':
            line = f.readline()
        if line == '':
            print 'Error: No sequences found in %s.' % inputFile
            exit(1)

        line = f.readline()
        # We are now at the first sequence
        sequences = []
        for i in range(Nstrands):
            LineList = WhiteSpaceSearch.split(line)
            sequences.append(LineList[2])
            line = f.readline()

        # Now blow through lines to number of bases
        while not(line[0].isdigit()) and line != '':
            line = f.readline()
        if line == '':
            print 'Error: No records found in input file %s.' % inputFile
            exit(1)

        # The line we're at now is the total number of bases
        LineList = WhiteSpaceSearch.split(line)
        nBases = int(LineList[0])

        # Set up variables
        fup = [0.0]*Nstrands
        StrandKey = []
        for i in range(Nstrands):
            for j in range(len(sequences[i])):
                StrandKey.append(i)

        # Now read records and find unpaired bases
        line = f.readline()
        while line != '' and line[0].isdigit():
            LineList = WhiteSpaceSearch.split(line)
            if int(LineList[1]) == nBases + 1:
                fup[StrandKey[int(LineList[0])-1]] += float(LineList[2])
            line = f.readline()
        f.close()

        for i in range(Nstrands):
            # fup[i] /= len(sequences[i])
            fup[i] *= x0[i]
            x0[i] *= len(sequences[i])

        fUnpaired[k] = sum(fup)/sum(x0)



# Make input file
f = open(outputFile,'w')
for i in range(len(Tlist)):
    f.write('%.1f\t%8.7f\n' % (Tlist[i],fUnpaired[i]))
f.close()


exit(0)

File: drawing-hg/meltplot/MeltParser.rb
#!/usr/bin/env ruby

# Parser of melt data, first hack version, written by JSB and MBP.

prefix = ARGV[0]
tempStart = (ARGV[1]).to_f
tempIncrement = (ARGV[2]).to_f
tempEnd = (ARGV[3]).to_f
outputFile = ARGV[4]
temperatures = Array.new
temp = tempStart
nTemps = 0
while temp <= tempEnd
  temperatures << temp
  temp += tempIncrement
  nTemps += 1
end #while

fUnpaired = Array.new # Fraction of srands unpaired

temperatures.each do |x|
  
  number_of_strands = 0
  n = 0 # Total strand length
  fUnpaired << 0.0
  
  inputFileName = sprintf("./%.1f/%s.fpairs",x.to_f,prefix)

  File.open(inputFileName) do |f|
    seen_id_seq = false
    f.each_line do |l|
      line = l.chomp

      # Fin
      if line =~ /% Number of strands: (\d+)/
        number_of_strands = $1.to_i
        next
      end #if

      # Check to see if we're at the line with % id sequence
      if line =~ /% id sequence/
        seen_id_seq = 1
        next
      end #if

      # Now we're parsing the sequences
      if seen_id_seq
        line =~ /%  #{seen_id_seq} (.+)$/
        sequence = $1
        n += sequence.length
        seen_id_seq += 1
        if seen_id_seq > number_of_strands
          seen_id_seq = false # go back to normal parsing
        end
        next
      end #if

      # Blast past line if it's a comment or newline
      next if line =~ /^%/

      # Split the line if it's got data
      chunks = l.chomp.split(/\s+/)

      if chunks[1].to_i == n+1  # If it's an entry for unpaired
        fUnpaired[fUnpaired.length - 1] += (chunks[2]).to_f
      end #if

    end #do
  end #do

  # Have to divide by n
  fUnpaired[fUnpaired.length - 1] /= n.to_f
  
end #do

# Write results to a file
outfile = File.open(outputFile,"w")
for i in 0..(nTemps-1)
  outfile.printf("%.1f\t%8.7f\n",temperatures[i],fUnpaired[i])
end #for
outfile.close


